src/starterbot/ProductionManager.cpp

```diff
@@ -60,6 +60,29 @@ void ProductionManager::onFrame() {
     // Release any workers that are idle, i.e. not constructing any buildings.
     m_unitManager.releaseUnits(m_workers, bw::Filter::IsIdle);
 
+    // Check to see if refinery exists
+    for (bw::Unit unit : m_unitManager.borrowUnits(bw::Filter::IsRefinery)) {
+        if (m_gas_gatherers.size() < 2) {
+            m_gas_gatherers = m_unitManager.reserveUnits(bw::Filter::IsWorker, 2);
```

Comment:
Subtle bug here. If `m_gas_gatherers` has one unit in it, then you reserve two new units, and the old unit is thrown away without being freed. So you've effectively got a unit that has disappeared from `UnitManager` and can never be used again.

You should probably do
```cpp
if (m_gas_gatherers.size() < 2) {
    bw::Unitset newGatherers = m_unitManager.reserveUnits(bw::Filter::IsWorker, 2 - m_gas_gatherers.size());
    m_gas_gatherers.insert(newGatherers.begin(), newGatherers.end());
}
```